src/providers/twitch/IrcMessageHandler.cpp

```diff
@@ -21,6 +21,101 @@
 
 namespace chatterino {
 
+static float relativeSimilarity(const QString &str1, const QString &str2)
+{
+    // Longest Common Substring Problem
+    std::vector<std::vector<int>> tree(str1.size(),
+                                       std::vector<int>(str2.size(), 0));
+    int z = 0;
+
+    for (int i = 0; i < str1.size(); ++i)
+    {
+        for (int j = 0; j < str2.size(); ++j)
+        {
+            if (str1[i] == str2[j])
+            {
+                if (i == 0 || j == 0)
+                {
+                    tree[i][j] = 1;
+                }
+                else
+                {
+                    tree[i][j] = tree[i - 1][j - 1] + 1;
+                }
+                if (tree[i][j] > z)
+                {
+                    z = tree[i][j];
+                }
+            }
+            else
+            {
+                tree[i][j] = 0;
+            }
+        }
+    }
+
+    return z == 0 ? 0.f : float(z) / std::max(str1.size(), str2.size());
+};
+
+float IrcMessageHandler::similarity(
+    MessagePtr msg, const LimitedQueueSnapshot<MessagePtr> &messages)
+{
+    float ret = 0.0f;
+    int bysameuser = 0;
+    for (int i = 1; bysameuser < getSettings()->hideSimilarMaxMessagesToCheck;
+         ++i)
+    {
+        if (messages.size() < i)
+        {
+            break;
+        }
+        const auto &prev_msg = messages[messages.size() - i];
+        if (prev_msg->parseTime.secsTo(QTime::currentTime()) >=
+            getSettings()->hideSimilarMaxDelay)
+        {
+            break;
+        }
+        if (msg->loginName != prev_msg->loginName)
+        {
+            continue;
+        }
+        ++bysameuser;
+        ret = std::max(
+            ret, relativeSimilarity(msg->messageText, prev_msg->messageText));
+    }
+    return ret;
+}
+
+float IrcMessageHandler::similarityRecentMessages(
+    const std::vector<MessagePtr> &messages, int pos_size)
+{
+    float ret = 0.0f;
+    int bysameuser = 0;
+    for (int i = 1; bysameuser < getSettings()->hideSimilarMaxMessagesToCheck;
+         ++i)
+    {
+        if (pos_size < i)
+        {
+            break;
+        }
+        const auto &prev_msg = messages[pos_size - i];
+        // is not able to time it, since parseTime equals currentTime during recent messages
+        if (prev_msg->parseTime.secsTo(QTime::currentTime()) >=
+            getSettings()->hideSimilarMaxDelay)
+        {
+            break;
+        }
+        if (messages[pos_size]->loginName != prev_msg->loginName)
+        {
+            continue;
+        }
+        ++bysameuser;
+        ret = std::max(ret, relativeSimilarity(messages[pos_size]->messageText,
+                                               prev_msg->messageText));
+    }
+    return ret;
+}
```

Comment:
To be honest I don't really see a point in trying to check the recent messages. Even if they are spam I don't feel like it's necessary to filter them.

Reply:
removed, will not check them since there's no immediate way to find their time